src/Storages/ObjectStorageQueue/ObjectStorageQueueMetadata.cpp

```diff
@@ -687,63 +686,145 @@ size_t ObjectStorageQueueMetadata::unregisterActive(const StorageID & storage_id
     const auto self = Info::create(storage_id);
 
     Coordination::Error code = Coordination::Error::ZOK;
+
     for (size_t i = 0; i < 1000; ++i)
     {
-        Coordination::Stat stat;
-        std::string registry_str;
-        auto zk_client = getZooKeeper();
+        Coordination::Requests requests;
+        Coordination::Responses responses;
+        size_t count = 0;
 
-        bool node_exists = zk_client->tryGet(registry_path, registry_str, &stat);
-        if (!node_exists)
+        bool supports_remove_recursive = true;
+        zkutil::ZooKeeperPtr zk_client;
+
+        try
         {
-            LOG_WARNING(log, "Cannot unregister: registry does not exist");
-            chassert(false);
-            return 0;
-        }
+            zk_client = getZooKeeper();
+            supports_remove_recursive = zk_client->isFeatureEnabled(DB::KeeperFeatureFlag::REMOVE_RECURSIVE);
 
-        Strings registered;
-        splitInto<','>(registered, registry_str);
+            Coordination::Stat stat;
+            std::string registry_str;
+            bool node_exists = zk_client->tryGet(registry_path, registry_str, &stat);
+            if (!node_exists)
+            {
+                LOG_WARNING(log, "Cannot unregister: registry does not exist");
+                chassert(false);
+                return 0;
+            }
 
-        bool found = false;
-        std::string new_registry_str;
-        size_t count = 0;
-        for (const auto & elem : registered)
-        {
-            if (elem.empty())
-                continue;
+            Strings registered;
+            splitInto<','>(registered, registry_str);
+
+            bool found = false;
+            std::string new_registry_str;
+            for (const auto & elem : registered)
+            {
+                if (elem.empty())
+                    continue;
+
+                auto info = Info::deserialize(elem);
+                if (info == self)
+                    found = true;
+                else
+                {
+                    if (!new_registry_str.empty())
+                        new_registry_str += ",";
+                    new_registry_str += elem;
+                    count += 1;
+                }
+            }
+            if (!found)
+                throw Exception(ErrorCodes::LOGICAL_ERROR, "Cannot unregister: table '{}' is not registered", self.table_id);
 
-            auto info = Info::deserialize(elem);
-            if (info == self)
-                found = true;
+            if (remove_metadata_if_no_registered && count == 0)
+            {
+                LOG_TRACE(log, "Removing all metadata in keeper by path: {}", zookeeper_path.string());
+                if (supports_remove_recursive)
+                {
+                    requests.push_back(zkutil::makeCheckRequest(registry_path, stat.version));
+                    requests.push_back(zkutil::makeRemoveRecursiveRequest(zookeeper_path, -1));
```

Comment:
The main fix of this PR is here: `makeRemoveRecursiveRequest` which removes `zookeeper_path` (e.g. all s3queue metadata in keeper) in the same keeper transaction with check for `registry_path`'s version. Because another table could try to register concurrently and before this change there would be race between checking metadata `count` (ref count), seeing it is equal to `0` and removing all metadata recursively while some other table could have registered in the middle.